docs/architecture/adr-038-add-support-for-wasm-light-clients.md

```diff
@@ -0,0 +1,86 @@
+# ADR 38: Add support for Wasm based light client
+
+## Changelog
+
+- 26/11/2020: Initial Draft
+
+## Status
+
+Draft
+
+## Abstract
+
+Currently in the SDK, light clients are part of Cosmos SDK codebase. What this means that, anytime a blockchain built 
+upon Cosmos SDK need to add support for new light client, Cosmos SDK need to be modified and all validator nodes of that 
+blockchain need to be updated with custom version of Cosmos SDK. 
+
+To remedy these shortcomings, we are proposing a WASM VM to host light client bytecode, which allows easier upgrading of 
+existing light clients as well as adding support for new light clients without requiring a code release and corresponding 
+hard-fork event.
+
+## Context
+Currently in the SDK, light clients are defined as part of the codebase and are implemented as submodules under 
+`/x/ibc/light-clients/`. To add support for new light client or update an existing light client in the event of security 
+issue or consensus update, we need to modify the codebase and integrate it in *numerous* places. In addition, individual 
```

Comment:
This is not true in regards to adding new light clients. You just need to register the proto definition on the codec. We refactored core IBC to allow for modular client registration. Light clients do not need to live under `x/ibc/light-clients` to be added to a blockchain.

As for existing light clients, the code in its sub module just needs to be modified

Reply:
You still need to manually upgrade the chain. The wasm based light client will allow you to add new client implementations without upgrading.

Reply:
as @AdityaSripal put down below
> Adding new light client implementations require governance support and is expensive: This is not ideal as chain governance is gatekeeper for new light client implementations getting added. If a small community want support for light client X, they may not be able to convince governance to support it. Making adding new light client implementations permissionless is a better solution.

My comments in the context section are derived from a mis-framing of the problem. The problem here is not writing/modifying code, but actually getting a chain to agree to run that code. I think re-framing this context would help with understanding the benefits of the wasm IBC light client

Reply:
Yes, this should be framed in terms of operating chains, not SDK codebase details, which are not directly relevant.

Reply:
Context is now framed in terms of operating chains.